[ATOM-2430] Better Error Message For Unknown SRG Field

galibzon opened the pull request:
With this fix, AZSLc fails when undefined SRG variables or functions
are detected during code emission.

BEFORE:
The shader pipeline would rely on DXC to detect the error,
which would be a non-informative error like:
"""
shadersource:561:23: error: use of undeclared identifier 'MaterialSrg'"
"""

NOW:
AZSLc itself detects the error and the message looks like:
"""
StandardPBR_ForwardPass.azslin(1031,24) : Emitter error #267:
Undefined ShaderResourceGroup symbol 'MaterialSrg::m_specularF0Factor'
"""

Added a couple of test cases where the error is expected to occur.

Signed-off-by: garrieta <garrieta@amazon.com>

Commit:
[ATOM-2430] Better Error Message For Unknown SRG Field

This is version 1.7.27

With this fix, AZSLc fails when undefined SRG variables or functions
are detected during code emission.

BEFORE:
The shader pipeline would rely on DXC to detect the error,
which would be a non-informative error like:
"""
shadersource:561:23: error: use of undeclared identifier 'MaterialSrg'"
"""

NOW:
AZSLc itself detects the error and the message looks like:
"""
StandardPBR_ForwardPass.azslin(1031,24) : Emitter error #267:
Undefined ShaderResourceGroup symbol 'MaterialSrg::m_specularF0Factor'
"""

Added a couple of test cases where the error is expected to occur.

Signed-off-by: garrieta <garrieta@amazon.com>

santorac commented on src/AzslcEmitter.cpp:
```diff
@@ -1211,6 +1211,34 @@ namespace AZ::ShaderCompiler
         }
     }
 
+    void CodeEmitter::ValidateIsNotUndefinedSrgVariableOrThrow(antlr4::Token* token, TokenToAst::AstNode* nodeFromToken) const
```
That's pretty verbose, hehe. I think "ValidateSrgMemberIsDefined" would be sufficient :)

galibzon replied:
Not sure why my brain chose NotUndefined... Horrible. Thanks for the tip.

galibzon replied:
Actually, I recall why. 

The idea is that not only SRG symbols will go through that function. So technically it checks that it is Not and undefined Srg Symbol. Nonetheless it can be confusing.

santorac commented on src/AzslcEmitter.cpp:
```diff
@@ -1211,6 +1211,34 @@ namespace AZ::ShaderCompiler
+                const auto& [uid, kindInfo] = *idAndKind;
+                if (kindInfo.IsKindOneOf(Kind::ShaderResourceGroup))
+                {
+                    const string errorMessage = FormatString("\nUndefined ShaderResourceGroup symbol '%s'", qualifiedIdCtx->getText().c_str());
```
Why the extra "\n" at the beginning?

galibzon replied:
That's a useless left over. Will remove. Thanks for the catch.

santorac commented on src/AzslcEmitter.h:
```diff
@@ -189,6 +189,13 @@ namespace AZ::ShaderCompiler
+        //! It checks whether the token that is about to be written to the output
+        //! is not an undefined SRG field/variable.
+        //! Throws an exception if it is an undefined ShaderResourceGroup field/variable.
+        void ValidateIsNotUndefinedSrgVariableOrThrow(antlr4::Token* token, TokenToAst::AstNode* nodeFromToken) const;
```
Maybe use the tern "Member" instead of "Variable" since it could be a function right?

galibzon replied:
I was thinking Symbol, but Member works. Will change to Member.

Commit:
Removed leading "\n" in exception message

Renamed
CodeEmitter::ValidateIsNoUndefinedSrgVariableOrThrow
to
CodeEmitter::IfIsSrgMemberValidateIsDefined

Signed-off-by: garrieta <garrieta@amazon.com>

moudgils reviewed (approved):
Other than the points raised by Chris this PR looks good to me
